File: Models/PizzaPedido.cs
namespace pizzaria;
using System;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

public class PizzaPedido 
{
    [Key]
    public int  Id { get; set; }
    public List<Sabor> Sabores { get; set; } = new();
    public Tamanho Tamanho { get; set; }
    public double Preco { get; private set; }
    [JsonIgnore]
    public PedidoFinal? PedidoFinal { get; set; }

    public PizzaPedido() { }

    public PizzaPedido(List<Sabor> sabores, Tamanho tamanho)
    {
        Sabores = sabores;
        Tamanho = tamanho;
        CalcularPreco();
    }

    public double CalcularPreco()
    {
        // preco = ((sabor1 + sabor2...saborN) / quantidade de sabores) + (preco * tamanho)

        double precoTotal = 0.0;

        foreach(Sabor sabor in Sabores)
        {
            precoTotal += sabor.Preco;
        }

        if (Sabores.Count > 0)
        {
            precoTotal = precoTotal / Sabores.Count;
        }

        precoTotal += precoTotal * (Tamanho.MultiplicadorPreco / 100.0);

        Preco = precoTotal;
        return Preco;
    }

    override public string ToString()
    {
        Console.Write($"Sabores: ");
        for (int i = 0; i < Sabores.Count; i++)
        {
            Console.Write(Sabores[i].Nome);
            if (i != (Sabores.Count - 1))
                Console.Write(", ");
        }

        Console.Write($" | Tamanho: {Tamanho.Nome} | ");
        Console.Write($"Preço: R${Preco}");

        return "";
    }
}

File: Frontend/src/app/models/entities/PizzaPedido.ts
import { Sabor } from "./Sabor";
import { Tamanho } from "./Tamanho";
import {Pedido} from "../interfaces/Pedido";
import {CalcularPrecoPizza} from "../CalcularPrecoPizza";

export class PizzaPedido implements Pedido {
  id: number = 0;
  sabores: Sabor[] = [];
  tamanho: Tamanho = new Tamanho();
  quantidade: number = 1;

  constructor(sabores: Sabor[], tamanho: Tamanho, quantidade: number) {
    this.sabores = sabores;
    this.tamanho = tamanho;
    this.quantidade = quantidade;
  }

  get preco() {
    return CalcularPrecoPizza(this.sabores, this.tamanho, this.quantidade);
  }

  getDescricao(): string {
    const tamanho = `Pizza ${this.tamanho.nome} de `;
    let nomesSabores = "";

    for (let i = 0; i < this.sabores.length; i++) {
      if (i == this.sabores.length - 1) {
        nomesSabores = nomesSabores + this.sabores[i].nome;
        break;
      }
      nomesSabores = nomesSabores + this.sabores[i].nome + ", ";
    }

    return tamanho + nomesSabores;
  }

  toJSON() {
    return {
      id: this.id,
      sabores: this.sabores,
      tamanho: this.tamanho,
      quantidade: this.quantidade,
      preco: this.preco
    }
  }
}
